src/naive/date.rs

```diff
@@ -782,9 +782,15 @@ impl NaiveDate {
 
     /// Add a duration of `i32` days to the date.
     pub(crate) const fn add_days(self, days: i32) -> Option<Self> {
-        if days == 0 {
-            return Some(self);
+        // fast path if the result is within the same year
+        const ORDINAL_MASK: i32 = 0b1_1111_1111_0000;
+        if let Some(ordinal) = ((self.ymdf & ORDINAL_MASK) >> 4).checked_add(days) {
```

Comment:
Does this have good test coverage?

Reply:
I'll have a look later today

Reply:
Here is the [current test of `add_days`](https://github.com/chronotope/chrono/blob/v0.4.26/src/naive/date.rs#L2785).

This is such a crucial function that I would take the time to test it thoroughly.

I recommend adding a `#[cfg(test)]`-only code path within the `add_days` to turn on and turn off the fast path. There's probably many tricks to do this. One way is to have a magic file that is checked. This file can be created or deleted by the test case code.

```rust
        // fast path if the result is within the same year
        const ORDINAL_MASK: i32 = 0b1_1111_1111_0000;

        #[cfg(test)]
        let mut ADD_DAYS_ENABLE_FAST_PATH: bool = std::path::Path("/tmp/ADD_DAYS_ENABLE_FAST_PATH").is_file();
        #[not(cfg(test))]
        const ADD_DAYS_ENABLE_FAST_PATH: bool = false;

        if ADD_DAYS_ENABLE_FAST_PATH && let Some(ordinal) = ((self.ymdf & ORDINAL_MASK) >> 4).checked_add(days) {
           // ...
```

The test case code would create and later delete the file `"/tmp/ADD_DAYS_ENABLE_FAST_PATH"` as needed.

I recommend test cases for `days` values
* `0` to `367` that compares `add_days` when `ADD_DAYS_ENABLE_FAST_PATH` is `true` and `false`

(this "turn on / turn off" trick could also be used in the bench so future bench runners can reprove to themselves that the function is still worthwhile).

<br/>

I recommend moving that one crucial operation `let Some(ordinal) = ((self.ymdf & ORDINAL_MASK) >> 4).checked_add(days)` out to it's own private _testable_ function and then test that function for all values from `0` to `366` inclusive. Prove beyond a doubt that statement is behaving as expected. It's fairly straightforward to do and adds needed confidence for this change.